src/resource-manager-session.c

```diff
@@ -135,3 +135,46 @@ save_session (ResourceManager *resmgr,
     g_clear_object (&cmd);
     return resp;
 }
+/*
+ * This function will send the required commands in order to bring a
+ * session back into the gap interval. This means loading and saving
+ * the associated context. If the SessionEntry is not saved then this
+ * function is a no-op.
+ */
+gboolean
+regap_session (ResourceManager *resmgr,
+               SessionEntry *entry)
+{
+    SessionEntryStateEnum state;
+    Tpm2Response *resp = NULL;
+    gboolean ret = TRUE;
+
+    g_assert_nonnull (resmgr);
+    g_assert_nonnull (entry);
+
+    state = session_entry_get_state (entry);
+    g_debug ("%s: swapping SessionEntry 0x%" PRIxPTR " in state \"%s\"",
+             __func__, (uintptr_t)entry, session_entry_state_to_str (state));
+    if (state == SESSION_ENTRY_SAVED_CLIENT ||
+        state == SESSION_ENTRY_SAVED_CLIENT_CLOSED ||
+        state == SESSION_ENTRY_SAVED_RM)
+    {
+        resp = load_session (resmgr, entry);
+        if (tpm2_response_get_code (resp) != TSS2_RC_SUCCESS) {
+            g_critical ("%s: Failed to save SessionEntry 0x%" PRIxPTR "removing "
+                        "from list", __func__, (uintptr_t)entry);
+            flush_session (resmgr, entry);
+            ret = FALSE;
+        }
+        resp = save_session (resmgr, entry);
```

Comment:
Are we not potentially trying to call `save_session()` on a session for which `load_session()` failed? Are there any negative consequences of that?

Reply:
Good catch. This function was missing a `goto` in that first error handling block.